master deployment

- [x] proxy setup for backend
- [x] Ansible setup to install PostgreSQL etc. (is already installed)
- [x] build&deploy [ci/Jenkinsfile](https://github.com/ElektraInitiative/PermaplanT/blob/master/ci/Jenkinsfile) (cargo run, diesel setup, diesel migration run)
- [x] make Jenkins job public (or at least accessible for PermaplanT users)
- [ ] (database) backup

Hello,

next update on the infrastructure :rocket:

Dev uses the latest rust version via rustup installed to `/opt/rust`. A `source`-able env script is provided under `/opt/rust.env`. Because the `rust` version provided by the os package manager is too old (of course).

I've tried to build the latest master on dev, but unfortunately I got errors like:

```
error: Table contains more than 32 columns. Consider enabling the `64-column-tables` feature to enable diesels support for tables with more than 32 columns.
error[E0433]: failed to resolve: use of undeclared crate or module `plant_detail`
```

@aidnurs @buenaflor Do you have any idea on what I could have done wrong?

`backend/README.md` states the following commands:

```
cargo install diesel_cli --no-default-features --features postgres
LC_ALL=C ~/.cargo/bin/diesel setup # needs db
LC_ALL=C ~/.cargo/bin/diesel migration run # needs db
cargo install typeshare-cli
cargo build
cargo run
```

Unfortunately I need to issue the `diesel` commands for the schema.rs to be written, but `diesel` already requires a functioning Postgres database. In the build phase I don't want to connect to any database. Connecting to the database is ok for testing and for the development server, but not in the build step as it should just throw out the binary.

Questions for the meeting on 2022-02-17:

1. Should we pin the rust version in any way, or is it ok to always use the latest version?
2. How should we handle the situation during build? (no Postgres, etc.)
3. How do we inject the correct `BASE_URL` for the backend into the frontend?

--- a/ci/Jenkinsfile
+++ b/ci/Jenkinsfile
@@ -1,12 +1,12 @@
 pipeline {
     agent { label 'permaplant' }
 
-    environment {
-        DATABASE_URL = 'postgres://cidb:cidb@127.0.0.1/cidb'
-    }
-
     stages {
         stage('Build') {
+            environment {
+                DATABASE_URL = 'postgres://cidb:cidb@127.0.0.1/cidb'
+            }
+
             steps {
                 echo 'Resetting the CI DB ...'
                 sh 'sudo /usr/local/bin/permaplant-reset-ci.sh'
@@ -14,8 +14,6 @@ pipeline {
                 echo 'Building...'
                 sh './ci/build-scripts/build-backend.sh'
                 sh './ci/build-scripts/build-frontend.sh'
-                stash includes: 'backend/target/release/backend', name: 'backend'
-                stash includes: 'frontend/dist/**/*', name: 'frontend'
             }
         }
 
@@ -26,8 +24,8 @@ pipeline {
 
             steps {
                 echo 'Deploying to Dev ...'
-                unstash 'backend'
-                unstash 'frontend'
+                sh 'sudo -u permaplant /usr/local/bin/permaplant-setup-dev.sh'
+                sh 'sudo /usr/local/bin/permaplant-deploy-dev.sh'
             }
         }
     }

--- a/ci/Jenkinsfile
+++ b/ci/Jenkinsfile
@@ -8,6 +8,8 @@ pipeline {
             }
 
             steps {
+                echo "Workspace is '${WORKSPACE}'"
+
                 echo 'Resetting the CI DB ...'
                 sh 'sudo /usr/local/bin/permaplant-reset-ci.sh'
 

--- a/ci/Jenkinsfile
+++ b/ci/Jenkinsfile
@@ -2,7 +2,7 @@ pipeline {
     agent { label 'permaplant' }
 
     stages {
-        stage('Build') {
+        stage('Build PR') {
             environment {
                 DATABASE_URL = 'postgres://cidb:cidb@127.0.0.1/cidb'
             }
@@ -16,18 +16,20 @@ pipeline {
                 echo 'Building...'
                 sh './ci/build-scripts/build-backend.sh'
                 sh './ci/build-scripts/build-frontend.sh'
+
+                echo 'Deploying'
+                sh "sudo /usr/local/bin/permaplant-deploy.sh pr ${WORKSPACE}"
             }
         }
 
-        stage('Deploy to Dev') {
+        stage('Deploy Dev') {
             when {
                 branch "master"
             }
 
             steps {
                 echo 'Deploying to Dev ...'
-                sh 'sudo -u permaplant /usr/local/bin/permaplant-setup-dev.sh'
-                sh 'sudo /usr/local/bin/permaplant-deploy-dev.sh'
+                sh "sudo /usr/local/bin/permaplant-deploy.sh dev ${WORKSPACE}"
             }
         }
     }

--- a/ci/Jenkinsfile
+++ b/ci/Jenkinsfile
@@ -10,6 +10,9 @@ pipeline {
             steps {
                 echo "Workspace is '${WORKSPACE}'"
 
+                echo 'Ensure pr service is stopped'
+                sh 'sudo /usr/sbin/service permaplant-pr stop'
+
                 echo 'Resetting the CI DB ...'
                 sh 'sudo /usr/local/bin/permaplant-reset-ci.sh'
 

--- a/doc/ci.md
+++ b/doc/ci.md
@@ -1,11 +1,27 @@
 # CI/CD Pipeline
 
+## Restrictions
+
+The following restrictions apply:
+
+- We cannot build frontend and backend in parallel.
+  - frontend depends on backend (`typeshare` definition)
+  - we can only build them sequentially (`backend` > `typeshare` > `frontend`)
+- we cannot build without a database
+  - due to diesel
+
 ## Pull Requests
 
 Each and every pull request will be built and tested against a separate CI database.
 
 Steps for all PRs are:
 
-* prepare a new postgres database
+* prepare an empty Postgres database
 * call `./ci/build-scripts/build-backend.sh`, which builds the rust binary as well as the bindings' definition via `typeshare`
 * call `./ci/build-scripts/build-frontend.sh`, which does a full release build of the frontend
+
+It will be available for public access on [pr.permaplant.net](pr.permaplant.net).
+
+## Dev Environment
+
+The `master` branch will be automatically deployed to [dev.permaplant.net](dev.permaplant.net).

